File: examples/cross-lang/java/RunDemo.java
import com.sun.net.httpserver.HttpServer;
import io.uniflow.sdk.UniFlowClient;

import java.io.IOException;
import java.io.OutputStream;
import java.net.InetSocketAddress;
import java.nio.charset.StandardCharsets;
import java.nio.file.Files;
import java.nio.file.Path;

/**
 * Minimal Java demo (default package for easy javac from repo root).
 *
 *   javac -d out sdk/java/src/main/java/io/uniflow/sdk/*.java examples/cross-lang/java/RunDemo.java
 *   java -cp out RunDemo
 */
public class RunDemo {
    public static void main(String[] args) throws Exception {
        String orch = env("UNIFLOW_URL", "http://127.0.0.1:8787");
        int greeterPort = Integer.parseInt(env("GREETER_PORT", "9101"));
        String greeterUrl = env("GREETER_URL", "http://127.0.0.1:" + greeterPort + "/execute");

        HttpServer greeter = null;
        if (!"1".equals(System.getenv("SKIP_LOCAL_GREETER"))) {
            greeter = startGreeter(greeterPort);
        }

        Path yaml = Path.of("examples/cross-lang/greeter.workflow.yaml");
        if (!Files.isRegularFile(yaml)) {
            yaml = Path.of("greeter.workflow.yaml");
        }

        UniFlowClient client = new UniFlowClient(orch);
        var v = client.validate(yaml.toString());
        v.raiseForStatus();
        System.out.println("validate ok: " + v.workflowId);

        String bindings = "{\"demo.greeter\":{\"type\":\"http\",\"endpoint\":\"" + greeterUrl + "\"}}";
        String reg = client.loadAndRegister(yaml.toString(), bindings);
        System.out.println("registered: " + reg);

        String run = client.startWorkflow("cross-lang-greeter", "{\"task\":\"java\"}", true);
        System.out.println("run: " + run);

        if (greeter != null) {
            greeter.stop(0);
        }
    }

    private static HttpServer startGreeter(int port) throws IOException {
        HttpServer server = HttpServer.create(new InetSocketAddress("127.0.0.1", port), 0);
        server.createContext("/execute", exchange -> {
            byte[] body = exchange.getRequestBody().readAllBytes();
            String task = "world";
            String raw = new String(body, StandardCharsets.UTF_8);
            int idx = raw.indexOf("\"task\"");
            if (idx >= 0) {
                int colon = raw.indexOf(':', idx);
                int q1 = raw.indexOf('"', colon + 1);
                int q2 = raw.indexOf('"', q1 + 1);
                if (q1 >= 0 && q2 > q1) task = raw.substring(q1 + 1, q2);
            }
            byte[] resp = ("{\"content\":\"hello " + task + "\",\"toolCalls\":[],\"stopReason\":\"stop\",\"metadata\":{}}")
                    .getBytes(StandardCharsets.UTF_8);
            exchange.getResponseHeaders().add("Content-Type", "application/json");
            exchange.sendResponseHeaders(200, resp.length);
            try (OutputStream os = exchange.getResponseBody()) {
                os.write(resp);
            }
        });
        server.start();
        return server;
    }

    private static String env(String k, String d) {
        String v = System.getenv(k);
        return v == null || v.isBlank() ? d : v;
    }
}

File: examples/cross-lang/python/run_demo.py
"""
Cross-lang Python demo: assume greeter Unit + Orchestrator already running,
or start greeter here and talk to Orchestrator.

Usage (from repo root, after pip install -e sdk/python and Orchestrator up):

  set UNIFLOW_URL=http://127.0.0.1:8787
  set GREETER_URL=http://127.0.0.1:9101/execute
  python examples/cross-lang/python/run_demo.py
"""

from __future__ import annotations

import json
import os
import sys
import threading
from http.server import BaseHTTPRequestHandler, HTTPServer
from pathlib import Path

ROOT = Path(__file__).resolve().parents[3]
sys.path.insert(0, str(ROOT / "sdk" / "python"))

from uniflow_sdk import UniFlowHttpClient  # noqa: E402

YAML_PATH = Path(__file__).resolve().parents[1] / "greeter.workflow.yaml"


class Greeter(BaseHTTPRequestHandler):
    def do_POST(self) -> None:  # noqa: N802
        length = int(self.headers.get("Content-Length", "0"))
        raw = self.rfile.read(length)
        task = "world"
        try:
            body = json.loads(raw.decode("utf-8"))
            task = body.get("input", {}).get("task", task)
        except Exception:  # noqa: BLE001
            pass
        payload = json.dumps(
            {"content": f"hello {task}", "toolCalls": [], "stopReason": "stop", "metadata": {}}
        ).encode("utf-8")
        self.send_response(200)
        self.send_header("Content-Type", "application/json")
        self.send_header("Content-Length", str(len(payload)))
        self.end_headers()
        self.wfile.write(payload)

    def log_message(self, fmt: str, *args) -> None:  # noqa: A003
        return


def maybe_start_greeter(port: int) -> HTTPServer | None:
    if os.environ.get("SKIP_LOCAL_GREETER") == "1":
        return None
    server = HTTPServer(("127.0.0.1", port), Greeter)
    threading.Thread(target=server.serve_forever, daemon=True).start()
    return server


def main() -> None:
    orch = os.environ.get("UNIFLOW_URL", "http://127.0.0.1:8787")
    greeter_port = int(os.environ.get("GREETER_PORT", "9101"))
    greeter_url = os.environ.get("GREETER_URL", f"http://127.0.0.1:{greeter_port}/execute")

    server = maybe_start_greeter(greeter_port)
    client = UniFlowHttpClient(orch)

    v = client.validate(YAML_PATH)
    v.raise_for_status()
    print("validate ok:", v.workflow_id)

    reg = client.load_and_register(
        YAML_PATH,
        bindings={"demo.greeter": {"type": "http", "endpoint": greeter_url}},
    )
    print("registered:", reg)

    run = client.start_workflow(reg["workflowId"], {"task": "python"}, sync=True)
    print("status:", run.get("status"))
    state = (run.get("result") or {}).get("state") or {}
    print("output.greet:", state.get("output.greet"))

    if server:
        server.shutdown()


if __name__ == "__main__":
    main()
